### PR TITLE
STRWEB-49: Support transpiling all modules defined in stripes config

### PR DESCRIPTION
**Pull request by mkuklis**
https://issues.folio.org/browse/STRWEB-49

@MikeTaylor this is a first early draft that allows for passing stripes config modules into the `babel-loader`.  The PR seems big but the most important part is here:

https://github.com/folio-org/stripes-webpack/compare/STRWEB-49?expand=1#diff-9defc5f35f4e8a9ffe6ec5567a6a76f929db7ee96140b1707e8a984f32360952R28-R34

This should allow for transpiring all modules with different namespaces (folio, reshare, etc) defined in stripes config (under modules).

### CODE REVIEW

**Comment by MikeTaylor** [comment]
Well, @mkuklis, this looks exactly right to me. I admit I don't follow what all the other parts of the PR are doing, but the change in `babelLoaderTest` seems just about perfect. My only reservation is over the test `fileName.includes(moduleName))`. That means that if I have a Stripes module `@mike/foo` and a non-Stripes Node package `@mike/foobar`, then the latter will get transpiled as well as the former. Can we easily make a stricter test?

**Comment by mkuklis** [comment]
@MikeTaylor the `moduleName` is actually misleading here. It should be `modulePath` because that's what we are doing here:

https://github.com/folio-org/stripes-webpack/pull/63/files#diff-9cab105d1391c8bfb52d9875539c54eb15493ad4a0edb2bbcc98c5a4bdbef1dfR96-R110

That should be sufficient to make sure other parts do not get in.

This PR still doesn't solve the issue when something is listed under given module as a dependency. For example:

https://github.com/openlibraryenvironment/ui-update/blob/master/package.json#L19

and I don't think we will be able to address it because you may have a mix of 3rd party dependencies (some maybe be already transpiled and some not). I think the only true solution for this is to make sure all consumed dependencies are actually pre-transpiled before they are published.

**Comment by MikeTaylor** [comment]
> @MikeTaylor the `moduleName` is actually misleading here. It should be `modulePath` because that's what we are doing here:
> 
> https://github.com/folio-org/stripes-webpack/pull/63/files#diff-9cab105d1391c8bfb52d9875539c54eb15493ad4a0edb2bbcc98c5a4bdbef1dfR96-R110

Ah yes. Looks like it should be fine unless I made a module called `@mike/foo/package.jsonbar`. And if I do that (and if Node doesn't straight up prevent me), then I'll only be getting what I richly deserve.

> This PR still doesn't solve the issue when something is listed under given module as a dependency. For example:
> 
> https://github.com/openlibraryenvironment/ui-update/blob/master/package.json#L19
> 
> and I don't think we will be able to address it because you may have a mix of 3rd party dependencies (some maybe be already transpiled and some not). I think the only true solution for this is to make sure all consumed dependencies are actually pre-transpiled before they are published.

Oh yes, I meant to reply to that bit. We already have a solution, thanks to Jason: `stripesDeps`. You can see it in use at https://github.com/openlibraryenvironment/ui-rs/blob/master/package.json#L60-L65

(`@reshare/update` should be specifying that `@reshare/stripes-reshare` is a stripes-dep. I'm guessing it gets away with it because it's always till now been used in a platform that also has `@reshare/rs', but I bet if you built it alone it would barf.)

**Comment by mkuklis** [comment]
@MikeTaylor this will be doable with `stripesDeps`

https://github.com/openlibraryenvironment/ui-rs/blob/master/package.json#L60

but I'm not sure if this is a good idea :(

I would rather provide pre-transpiled versions of these modules.

but we don't have that in place yet so perhaps this is a way to go initially.

**Comment by MikeTaylor** [comment]
I would be interested to hear why `stripesDeps` is not a good idea. But at any rate, it is what we have now and what we need to co-exist with. (Perhaps putting the `stripesDeps` up at the platform level would be better than having the individual modules state it.)

Anyway, let's not allow it to stop us fixing the present problem.

**Comment by mkuklis** [comment]
@MikeTaylor yes and since we have access to that package.json already it should not be hard to add. Let me take a look at this. I think we will end up with something workable at the end.

**Comment by MikeTaylor** [comment]
Is there any reason not to merge this PR?

**Comment by MikeTaylor** [comment]
Ah, I'd not realised that wasn't included already. Thanks for keeping this going!

**Comment by mkuklis** [comment]
@MikeTaylor this should now also support `stripesDeps`.

I tried different scenarios running these changes via yarn workspaces and via a single ui-module (https://github.com/openlibraryenvironment/ui-rs) with a config similar to:

````js
module.exports = {
  okapi: {
    url: 'https://folio-snapshot-okapi.dev.folio.org', tenant: 'diku',
  },
  config: {
    logCategories: 'core,path,action,xhr',
    logPrefix: '--',
    showPerms: false,
    hasAllPerms: false,
    languages: ['en'],
    suppressIntlErrors: true,
  },
  modules: {
    '@reshare/rs': {}
  },
};
````

since ui-rs has some `stripesDeps` listed in package.json:

https://github.com/openlibraryenvironment/ui-rs/blob/ad2e6353c569a6a0e89520cd04b8617ddf5fb0c9/package.json#L60-L65

I would appreciate if you could give it a try. The changes seem big but it really is just a single function at the end:

https://github.com/folio-org/stripes-webpack/pull/63/files#diff-9cab105d1391c8bfb52d9875539c54eb15493ad4a0edb2bbcc98c5a4bdbef1dfR154

**Comment by MikeTaylor** [comment]
> I would appreciate if you could give it a try.

I can see how to do that in workspace (by putting my git checkout of stripes-webpack branch `STRWEB-49` into the workspace alongside my apps), but now how to do it for a single app. Do you just manually modify the `node_modules` are to put your checkout into it?

**Comment by mkuklis** [comment]
@MikeTaylor yes that's exactly what I did. I just copied the stripes-webpack from STRWEB-49 branch into  the `node_modules` (under `@folio`).

**Comment by zburke** [comment]
@mkuklis, @MikeTaylor : 

> I can see how to [give it a try] in workspace (by putting my git checkout of stripes-webpack branch STRWEB-49 into the workspace alongside my apps), but now how to do it for a single app. 

Dependency version-values don't have to use semver; they can also be [local filesystem paths or git URLs or GitHub shortcuts](https://docs.npmjs.com/cli/v8/commands/npm-install). To test this branch in an individual app with only a transitive dep on this repository, use a `resolutions` entry in its `package.json` file:
```
resolutions: {
  "@folio/stripes-webpack": "folio-org/stripes-webpack#STRWEB-49"
}
```

As to "Is there any reason not to merge this PR?" I think the answer is, "Yes: we're at the beginning of a release window for 70+ apps so this is the wrong time to make fundamental changes to this tool."

**Comment by mkuklis** [comment]
@MikeTaylor you should be able to run a single module (I tried https://github.com/openlibraryenvironment/ui-rs)

by cloning a fresh version and adding:

````js
resolutions: {
  "@folio/stripes-webpack": "folio-org/stripes-webpack#STRWEB-49"
}
````

to the `package.json` as @zburke suggested above.

This seems to work for me after I added a reshare registry: 

`yarn config set @reshare:registry https://nexus.libsdev.k-int.com/repository/libsdev-npm-group`

(thanks for pointing it out @skomorokh!).

**Review by MikeTaylor — APPROVED** [review]
Yikes, how on earth did I allow this to fall through the gaps for so long? My apologies, @mkuklis. I have now tested this with my own non-`@folio` module and verified that it does what we need. (I have also learned a valuable trick along the way, depending on `"@folio/stripes-webpack": "folio-org/stripes-webpack#STRWEB-49"`.)

Thank you for working on this. Please go ahead an merge as soon as convenient!

**Comment by zburke** [comment]
@mkuklis, @MikeTaylor: MG is now pushed out five weeks with a GA date in late September. About [20% of the open tickets are in UI projects](https://issues.folio.org/issues/?jql=issuetype%20in%20(Bug%2C%20Story%2C%20Task)%20AND%20Release%20in%20(%22Morning%20Glory%20R2%202022%20Bug%20Fix%22)%20AND%20status%20not%20in%20(closed%2C%20%22in%20bugfix%20review%22%2C%20%22Awaiting%20deployment%22)%20ORDER%20BY%20key%20ASC). Do you think we should continue to sit on this PR until after MG releases are complete, or publish it now?

**Comment by MikeTaylor** [comment]
Realistically, there is no great rush from my point of view. So I reluctantly conclude that the prudent strategy is to wait for the delayed Morning Glory release before merging this.

But while it's not urgent, it's important -- so let's not let this slip through the cracks!

**Comment by MikeTaylor** [comment]
I guess now that [Morning Glory is out](https://folio-project.slack.com/archives/C1EPYQKCL/p1664224908593639) this can finally be merged?

**Comment by zburke** [comment]
I agree, @MikeTaylor. What do you think, @mkuklis? I think the sooner the better since Nolana is already staring at us.

**Comment by mkuklis** [comment]
@MikeTaylor and @zburke I just merged this in. Let's see if it causes any issues.

**Comment by MikeTaylor** [comment]
Woo-hoo!

Thanks again for all your work on this. Hopefully this is the last we'll see of specially privileging the `@folio` namespace — one small but important plank in the democratization of FOLIO module development!

**Pull request merged**
